### DISCUSSION TITLE
How to save output in a txt file

### DISCUSSION THREAD

**Discussion by gauravdesale8** [ultralytics/ultralytics#9217] [Q&A]
I have trained a best.pt file for detecting digits from water meters. When I make predictions, I want to save the detected digits in a text file. Additionally, every time I run the program, it should append the readings to the same text file. Please suggest a solution or workaround.

**Comment by pderrenger** [comment]
Hi there 👋! For saving the detected digits in a text file and appending new readings every time you run your program, you can use the `--save-txt` and `--name` flags in your prediction command. Here's a simple way to do it:

```bash
yolo detect predict model=best.pt source=path/to/your/images --save-txt --name my_results
```

This will save the detections in a `.txt` file inside `runs/detect/my_results/labels/` folder, with each detection in its own file named after the image. To append new readings, just keep the `--name my_results` consistent across runs.

Remember, the text files will contain the bounding box coordinates and class IDs, not the class names directly. If you're detecting digits, you'll probably have class IDs corresponding to each digit.

If you need further customization on how the data is saved (e.g., saving in a single, cumulative file or formatting the saved content), you might need a small script to post-process the output `.txt` files according to your needs. Let me know if you need help with that! 

Happy detecting! 😊

**Reply by gauravdesale8** [reply]
Okay, so for example:

The digits in my meter are "00072."
Can I write a function that will save my digits, "Date," "Time," and "00072"?
I want to actually save them in a CSV so that I can use them further for analysis purposes.
I did try to find some solutions on the internet, but the majority of content is just based on how to train the model. Can I make further modifications for additional requirements?

**Reply by gauravdesale8** [reply]
And also, the save_txt file is saving the digits randomly with coordinates. How can I avoid this? Can I implement a mechanism where it scans from left to right? So that my digits would look like this: "00072" not "20700"(with coordinates)

**Reply by pderrenger** [reply]
Hey there! 😊

To ensure your detected digits are saved in the order they appear from left to right, you might need a custom post-processing script. Here's a quick example in Python that sorts the detections based on their x-coordinate before saving:

```python
def save_sorted_detections(file_path, detections):
    # Sort detections by the x-coordinate of their bounding box
    sorted_detections = sorted(detections, key=lambda x: x[1])
    
    with open(file_path, 'a') as f:  # Use 'a' to append to an existing file
        for detection in sorted_detections:
            # Example format: ClassName, Date, Time, x1, y1, x2, y2, Confidence
            f.write(f'{detection[0]}, {detection[1]}, {detection[2]}, {detection[3]}\n')

# Sample usage
detections = [("2", "2023-04-12", "10:00", 100), ("0", "2023-04-12", "10:00", 50), ("7", "2023-04-12", "10:00", 150)]
save_sorted_detections('readings.csv', detections)
```

You'll need to adjust the script to match how you're obtaining the reading's characters from your detection output, especially the sorting key for the x-coordinate. This example assumes that `detections` is a list of tuples where each tuple is `(digit, date, time, x-coordinate)`.

Regarding saving digits without randomness, ensure your detections are ordered before running the script.

Hope this helps! Let me know if you have more questions. 🚀
